kernel/src/callback.rs

```diff
@@ -160,70 +158,123 @@ impl AppId {
+    ) -> Self {
+        Callback(Some(ProcessCallback::new(
+            app_id,
+            callback_id,
+            appdata,
+            fn_ptr,
+        )))
     }
 
-    pub fn schedule(&mut self, r0: usize, r1: usize, r2: usize) -> bool {
-        self.cb.map_or(true, |mut cb| cb.schedule(r0, r1, r2))
+    /// Get the contained [`ProcessCallback`] struct
+    ///
+    /// If the [`Callback`] refers to a process, this returns the
+    /// contained [`ProcessCallback`] struct.
+    pub(crate) fn into_inner(self) -> Option<ProcessCallback> {
+        self.0
     }
 
-    pub(crate) fn into_subscribe_success(self) -> GenericSyscallReturnValue {
-        match self.cb {
-            None => GenericSyscallReturnValue::SubscribeSuccess(0 as *mut u8, 0),
-            Some(cb) => GenericSyscallReturnValue::SubscribeSuccess(
-                cb.fn_ptr.as_ptr() as *const u8,
-                cb.appdata,
-            ),
-        }
+    /// Attempt to trigger the callback.
+    ///
+    /// If this callback refers to an actual function pointer (not a
+    /// _null callback_) and the process is valid, this will queue the
+    /// [`Callback`] for the associated process.
+    ///
+    /// ## Arguments
+    /// The arguments (`r0-r2`) are the values passed back to the
+    /// process and are specific to the individual `Driver`
+    /// interfaces.
+    ///
+    /// ## Return value
+    /// The following conditions are checked in this order to
+    /// determine the return value:
+    ///
+    /// 1. If this [`Callback`] is a default instance (does not refer
+    ///    to a process), it returns `true` and behaves as if the
+    ///    callback was scheduled.
+    ///
+    /// 2. This function returns `false` if the process is no longer
+    ///    active.
+    ///
+    /// 3. For a _null callback_, this function will return `true` and
+    ///    behave as if the callback was scheduled.
+    ///
+    /// 4. This function returns `false` if the queue for the process
+    ///    is full and the callback could not be scheduled.
+    ///
+    /// Otherwise, this function returns `true`.
+    pub fn schedule(&mut self, r0: usize, r1: usize, r2: usize) -> bool {
+        self.0.as_mut().map_or(true, |cb| cb.schedule(r0, r1, r2))
     }
+}
 
-    pub(crate) fn into_subscribe_failure(self, err: ErrorCode) -> GenericSyscallReturnValue {
-        match self.cb {
-            None => GenericSyscallReturnValue::SubscribeFailure(err, 0 as *mut u8, 0),
-            Some(cb) => GenericSyscallReturnValue::SubscribeFailure(
-                err,
-                cb.fn_ptr.as_ptr() as *const u8,
-                cb.appdata,
-            ),
-        }
+impl Default for Callback {
+    /// Construct a new default [`Callback`]
+    ///
+    /// A default [`Callback`] instance will not point to any actual
+    /// userspace process. Therefore, no actual callbacks will be
+    /// scheduled by this instance.
+    fn default() -> Self {
+        Callback(None)
     }
 }
```

Comment:
why manually implement default when the derived version would be the same?

Reply:
I'm implementing manually for the documentation comment (as this is a publicly exposed API), which will be visible in the generated rustdoc. I can switch to the derive macro if you want.

Reply:
no, that makes sense